x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java

```diff
@@ -381,19 +380,32 @@ public void deleteTrainedModel(String modelId, ActionListener<Boolean> listener)
 
     public void expandIds(String idExpression,
                           boolean allowNoResources,
-                          @Nullable PageParams pageParams,
+                          PageParams pageParams,
```

Comment:
So, are `pageParams` guaranteed to be non-null?

Reply:
Is the `pageParams` parameter really necessary? I'd argue that I would expect `expandIds` to return all the ids rather than a page as the id small compared to a full document. 

The default value of `PageParams.getSize()` is 100. Without setting a page param object you may realise that limitation is in place and that you are not getting all of your ids. Is there even a mechanism to say you need to get the second page now? 

Compare to `JobConfigProvider.expandJobIds`  which gives you `AnomalyDetectorsIndex.CONFIG_INDEX_MAX_RESULTS_WINDOW = 10_000` ids.

It would also simplify this code to remove to the pageParams which is a very good reason to remove it.

Reply:
> Is the `pageParams` parameter really necessary? 

Yes, I think it is. The APIs making calls against this function require paging params (default is from: 0, size: 100). I think this is a sane way of handling larger numbers of items.

> Compare to `JobConfigProvider.expandJobIds` which gives you `AnomalyDetectorsIndex.CONFIG_INDEX_MAX_RESULTS_WINDOW = 10_000` ids.

I think that this is a bug and shows that we don't support good pagination at all with anomaly detection jobs. 

>  Is there even a mechanism to say you need to get the second page now?

Yes, the total count of all the ids found with id pattern is returned.

Reply:
> So, are `pageParams` guaranteed to be non-null?

It is not possible to set them as `null` via the API since you can only set `from` and `size` individually.

If it is null from a client, they just don't set those parameters (and thus get the default value). 

If an internal caller is making these calls, they should not explicitly set pageParams to null and if they do, they will receive an error pretty quickly.

Reply:
My concern was that with default paging we open ourselves up to bugs such as elastic/kibana#38559.

I assumed `getTrainedModel` would take paging parameters in which case paging in `expandIds` would be redundant but I see from the way the code is used that is not the case and `GetTrainedModelsStatsAction` explicitly depends on paging ids. 

AD jobs are not paged because when they lived in clusterstate all where returned by the GET API. It would have been a breaking change to require paging on that API when we move to the .ml-config index and we could not make a breaking change in that release.

Reply:
Yes, the apis `GET _ml/trained_models` and `GET _ml/trained_models/_stats` accept paging parameters and use those when expanding the IDs. 

Since they are stored in indices, paging parameters are a must (even for expanding IDs) and might as well have it handled now instead of worrying about adding them later when we reach scaling issues.